From e21559aa22344ab54dd239caf6867f8f6dcb44ce Mon Sep 17 00:00:00 2001
From: Patrick Doyle <pdoyle@irobot.com>
Date: Wed, 13 Nov 2019 16:05:30 -0500
Subject: [PATCH] Always include root node when verifying image signatures.

If '/' is not included in the hashed-nodes property for the signature,
add it.
---
 common/image-sig.c | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/common/image-sig.c b/common/image-sig.c
index 004fbc5..88e81d4 100644
--- a/common/image-sig.c
+++ b/common/image-sig.c
@@ -361,14 +361,24 @@ int fit_config_check_sig(const void *fit, int noffset, int required_keynode,
 	/* Count the number of strings in the property */
 	prop = fdt_getprop(fit, noffset, "hashed-nodes", &prop_len);
 	end = prop ? prop + prop_len : prop;
-	for (name = prop, count = 0; name < end; name++)
+	bool root_not_included = true;
+	for (name = prop, count = 0; name < end; name++) {
 		if (!*name)
 			count++;
+		if (root_not_included		    &&
+		    (name == prop || name[-1] == 0) &&
+		    name[0] == '/'		    &&
+		    name[1] == 0) {
+			root_not_included = false;
+		}
+	}
 	if (!count) {
 		*err_msgp = "Can't get hashed-nodes property";
 		return -1;
 	}
-
+	if (root_not_included) {
+		count++;
+	}
 	if (prop && prop_len > 0 && prop[prop_len - 1] != '\0') {
 		*err_msgp = "hashed-nodes property must be null-terminated";
 		return -1;
@@ -384,7 +394,13 @@ int fit_config_check_sig(const void *fit, int noffset, int required_keynode,
 	char *node_inc[count];
 
 	debug("Hash nodes (%d):\n", count);
-	for (name = prop, i = 0; name < end; name += strlen(name) + 1, i++) {
+
+	if (root_not_included) {
+		node_inc[0] = "/";
+		debug("	  '/' (ADDED!)\n");
+	}
+
+	for (name = prop, i = root_not_included ? 1 : 0; name < end; name += strlen(name) + 1, i++) {
 		debug("   '%s'\n", name);
 		node_inc[i] = (char *)name;
 	}
-- 
2.7.4

